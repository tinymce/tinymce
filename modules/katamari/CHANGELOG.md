# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased

## 7.2.0 - 2021-05-06

### Added
- Introduced `Maybe` as an eventual replacement for `Optional`.
- Added `pipe` function to the `Fun` API.

### Changed
- Added type guard predicates to `Arr.find` methods #TINY-7138

## 7.1.0 - 2020-02-02

### Added
- New `Regex` module

## 7.0.0

### Removed
- Removed `Struct`. Please use TypeScript interfaces and functions instead.
