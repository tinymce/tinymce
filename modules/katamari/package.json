{
  "name": "@ephox/katamari",
  "version": "9.1.5",
  "description": "Basic data type library",
  "repository": {
    "type": "git",
    "url": "https://github.com/tinymce/tinymce.git",
    "directory": "modules/katamari"
  },
  "scripts": {
    "prepublishOnly": "tsc -b",
    "test": "bedrock-auto -b chrome-headless -d src/test/ts",
    "test-manual": "bedrock -d src/test/ts",
    "lint": "eslint --config ../../.eslintrc.json src/**/*.ts"
  },
  "author": "Ephox Corporation DBA Tiny Technologies, Inc",
  "license": "MIT",
  "dependencies": {
    "@ephox/dispute": "^1.0.3"
  },
  "files": [
    "lib/main",
    "lib/demo",
    "lib/test",
    "src",
    "tsconfig.json",
    "README.md",
    "LEGAL.txt",
    "CHANGELOG.md",
    "LICENSE.txt"
  ],
  "main": "./lib/main/ts/ephox/katamari/api/Main.js",
  "module": "./lib/main/ts/ephox/katamari/api/Main.js",
  "types": "./lib/main/ts/ephox/katamari/api/Main.d.ts"
}
