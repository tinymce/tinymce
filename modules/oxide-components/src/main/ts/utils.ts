import * as DefaultAvatar from './defaultavatar/DefaultAvatar';

export {
  DefaultAvatar
};
