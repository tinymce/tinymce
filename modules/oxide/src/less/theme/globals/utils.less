//
// Utils
//

// Luma checker
// Depending on @background-colorâ€™s luma returns either of two colors
// Usage example: @foo-color: .bg-luma-checker(@color-1, @color-2)[@result];
.bg-luma-checker(@color-light, @color-dark) {
  @result: if((luma(@background-color) > 50%), @color-light, @color-dark);
}
