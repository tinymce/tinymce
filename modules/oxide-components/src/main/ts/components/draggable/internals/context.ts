import { Fun } from '@ephox/katamari';
import { createContext, useContext } from 'react';

import type { DraggableState } from './types';

const initialState: DraggableState = { shift: { x: 0, y: 0 }, setShift: Fun.noop, draggableRef: { current: null }};
const DraggableContext = createContext<DraggableState>(initialState);

const useDraggable = (): DraggableState => {
  const context = useContext(DraggableContext);
  if (!context) {
    throw new Error('Draggable compound components must be rendered within the Draggable component');
  }
  return context;
};

export { useDraggable, DraggableContext };