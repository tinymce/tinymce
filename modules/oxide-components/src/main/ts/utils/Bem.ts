import { Arr, Obj, Strings, Type } from '@ephox/katamari';
import type { Classes } from '@tinymce/oxide/skins/ui/default/skin.ts';

type ValidClassName = keyof Classes;

type BlockElementModifier<K extends string> = K extends `${string}__${string}--${string}` ? K : never;
type BlockElement<K extends string> = K extends `${string}__${string}` ? (K extends `${string}--${string}` ? never : K) : never;
type BlockModifier<K extends string> = K extends `${string}--${string}` ? (K extends `${string}__${string}` ? never : K) : never;
type Block<K extends string> = K extends `${string}__${string}` | `${string}--${string}` ? never : K;

type BEMBlocks = Extract<ValidClassName, Block<ValidClassName>>;
type BEMBlockModifiers = Extract<ValidClassName, BlockModifier<ValidClassName>>;
type BEMBlockElements = Extract<ValidClassName, BlockElement<ValidClassName>>;
type BEMBlockElementModifiers = Extract<ValidClassName, BlockElementModifier<ValidClassName>>;

type SplitBlockElement<K extends string> = K extends `${infer Block}__${infer Element}` ? [Block, Element] : never;
type ValidElementPairs = SplitBlockElement<BEMBlockElements>;
type ElementForBlock<B extends BEMBlocks> = Extract<ValidElementPairs, [B, any]>[1];

type SplitBlockModifier<K extends string> = K extends `${infer Block}--${infer Modifier}` ? [Block, Modifier] : never;
type SplitBlockElementModifier<K extends string> = K extends `${infer Block}__${infer Element}--${infer Modifier}` ? [Block, Element, Modifier] : never;

type ValidBlockModifiers = SplitBlockModifier<BEMBlockModifiers>;
type ValidBlockElementModifiers = SplitBlockElementModifier<BEMBlockElementModifiers>;

type ModifiersForBlock<B extends BEMBlocks> = Extract<ValidBlockModifiers, [B, any]>[1];
type ModifiersForBlockElement<B extends BEMBlocks, E extends ElementForBlock<B>> = Extract<ValidBlockElementModifiers, [B, E, any]>[2];
type ModifierRecord<M extends string> = [M] extends [never] ? never : Partial<Record<M, boolean>>; // To avoid 'Record<never, boolean>' which is an empty object

const applyModifiers = <M extends string>(base: string, modifiers: ModifierRecord<M>): string => {
  const modifierClasses = Arr.filter(Obj.mapToArray(modifiers, (v, k) => v ? `${base}--${k}` : ''), Strings.isNotEmpty);
  return [ base, ...modifierClasses ].join(' ');
};

export const block = <B extends BEMBlocks, M extends ModifiersForBlock<B> = never>(
  block: B,
  modifiers?: ModifierRecord<M>
): string => Type.isUndefined(modifiers) ? block : applyModifiers(block, modifiers);

export const element = <B extends BEMBlocks, E extends ElementForBlock<B>, M extends ModifiersForBlockElement<B, E> = never>(
  block: B,
  element: E,
  modifiers?: ModifierRecord<M>
): string => {
  const base = `${block}__${element}`;
  return Type.isUndefined(modifiers) ? base : applyModifiers(base, modifiers);
};
