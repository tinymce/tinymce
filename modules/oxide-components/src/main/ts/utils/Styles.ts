import { type Classes } from '@tinymce/oxide/skins/ui/default/skin.ts';

export const classes = (classNames: Array<keyof Classes>): string => classNames.join(' ');
