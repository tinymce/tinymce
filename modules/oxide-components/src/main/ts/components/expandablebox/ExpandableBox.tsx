import { Type } from '@ephox/katamari';
import { useLayoutEffect, useRef, useState, type FC, type PropsWithChildren } from 'react';

import { Icon } from '../../internal/icon/Icon.component';
import * as Bem from '../../utils/Bem';

export interface ExpandableBoxProps extends PropsWithChildren {
  /** Icon resolver */
  iconResolver: (icon: string) => string;
  /** Expanded state */
  expanded?: boolean;
  /** Callback for toggle button */
  onToggle?: (curExpand: boolean) => void;
  /** Text to render when the button is to expand the content */
  expandText?: string;
  /** Text to render when the button is to collapse the content */
  collapseText?: string;
}

export const ExpandableBox: FC<ExpandableBoxProps> = ({
  iconResolver,
  expanded = false,
  onToggle,
  expandText = 'Expand',
  collapseText = 'Collapse',
  children
}) => {
  const contentRef = useRef<HTMLDivElement | null>(null);
  const [ overflowing, setOverflowing ] = useState(false);

  useLayoutEffect(() => {
    const contentEl = contentRef.current;
    if (Type.isNonNullable(contentEl)) {
      setOverflowing(contentEl.scrollHeight > contentEl.clientHeight);
    }
  }, [ children ]);

  return (
    <div className={Bem.block('tox-expandable-box')}>
      <div ref={contentRef} className={Bem.element('tox-expandable-box', 'content', { expanded, overflowing: overflowing && !expanded })}>
        {children}
      </div>
      {
        overflowing && <button type="button" className={Bem.element('tox-expandable-box', 'toggle-button')} onClick={() => onToggle?.(expanded)}>
          <Icon resolver={iconResolver} icon={expanded ? 'chevron-up' : 'chevron-down'} />
          {expanded ? collapseText : expandText}
        </button>
      }
    </div>
  );
};
