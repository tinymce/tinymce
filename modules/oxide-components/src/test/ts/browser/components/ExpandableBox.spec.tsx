import { Fun } from '@ephox/katamari';
import { userEvent, type Locator } from '@vitest/browser/context';
import { ExpandableBox, type ExpandableBoxProps } from 'oxide-components/components/ExpandableBox/ExpandableBox';
import * as Bem from 'oxide-components/utils/Bem';
import { useState } from 'react';
import { describe, expect, it } from 'vitest';
import { render } from 'vitest-browser-react';

describe('browser.ExpandableBoxTest', () => {
  const wrapper = ({ children }: { children: React.ReactNode }) => {
    return (
      <div className={Bem.block('tox')}>
        {children}
      </div>
    );
  };

  const defaultProps: ExpandableBoxProps = {
    expanded: false,
    onToggle: Fun.noop,
    maxHeight: 80,
    iconResolver: (icon: string) => `Icon: ${icon}`
  };

  const TestComponentToggle = (props: ExpandableBoxProps) => {
    const [ expanded, setExpanded ] = useState(false);

    return (
      <ExpandableBox
        {...props}
        expanded={expanded}
        onToggle={() => setExpanded(!expanded)}
      >
        {props.children}
      </ExpandableBox>
    );
  };

  const waitForElementText = async (getByText: (text: string) => Locator, text: string) => {
    await expect.poll(() => expect(getByText(text).element())).toBeTruthy();
  };

  it('Should be able to expand and collapse content that overflows', async () => {
    const TestComponent = () => {
      const [ expanded, setExpanded ] = useState(false);

      return (
        <ExpandableBox
          {...defaultProps}
          expanded={expanded}
          onToggle={() => setExpanded(!expanded)}
        >
          <div style={{ height: '200px' }}>Hello world</div>
        </ExpandableBox>
      );
    };

    const { asFragment, getByText } = render(<TestComponent />, { wrapper });

    await waitForElementText(getByText, 'Expand');
    expect(asFragment()).toMatchSnapshot('1. Before expand click');

    await userEvent.click(getByText('Expand'));
    await waitForElementText(getByText, 'Collapse');
    expect(asFragment()).toMatchSnapshot('2. After expand click');

    await userEvent.click(getByText('Collapse'));
    await waitForElementText(getByText, 'Expand');
    expect(asFragment()).toMatchSnapshot('3. After collapse click');
  });

  it('Should not render expand if the content fits within maxHeight', async () => {
    const { asFragment } = render(
      <ExpandableBox
        {...defaultProps}
      >
        <div style={{ height: '50px' }}>Hello world</div>
      </ExpandableBox>,
      { wrapper }
    );

    expect(asFragment()).toMatchSnapshot('Content fits within maxHeight');
  });

  it('Should be able to set maxHeight, collapseText and expandText', async () => {
    const { asFragment, getByText } = render(
      <TestComponentToggle
        {...defaultProps}
        maxHeight={100}
        expandText="Show more"
        collapseText="Show less"
      >
        <div style={{ height: '150px' }}>Hello world</div>
      </TestComponentToggle>,
      { wrapper }
    );

    expect(asFragment()).toMatchSnapshot('1. Content fits within maxHeight');

    await userEvent.click(getByText('Show more'));
    await waitForElementText(getByText, 'Show less');
    expect(asFragment()).toMatchSnapshot('2. After show mode click');
  });
});
