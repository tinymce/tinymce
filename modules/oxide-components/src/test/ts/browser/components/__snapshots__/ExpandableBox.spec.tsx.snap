// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`browser.ExpandableBoxTest > Should be able to expand and collapse content that overflows > 1. Before expand click 1`] = `
<DocumentFragment>
  <div
    class="tox"
  >
    <div
      class="tox-expandable-box"
    >
      <div
        class="tox-expandable-box__content tox-expandable-box__content--overflowing"
        style="max-height: 80px;"
      >
        <div
          style="height: 200px;"
        >
          Hello world
        </div>
      </div>
      <button
        class="tox-expandable-box__toggle-button"
        type="button"
      >
        <span
          class="tox-icon"
        >
          Icon: chevron-down
        </span>
        Expand
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`browser.ExpandableBoxTest > Should be able to expand and collapse content that overflows > 2. After expand click 1`] = `
<DocumentFragment>
  <div
    class="tox"
  >
    <div
      class="tox-expandable-box"
    >
      <div
        class="tox-expandable-box__content tox-expandable-box__content--expanded"
        style="max-height: none;"
      >
        <div
          style="height: 200px;"
        >
          Hello world
        </div>
      </div>
      <button
        class="tox-expandable-box__toggle-button"
        type="button"
      >
        <span
          class="tox-icon"
        >
          Icon: chevron-up
        </span>
        Collapse
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`browser.ExpandableBoxTest > Should be able to expand and collapse content that overflows > 3. After collapse click 1`] = `
<DocumentFragment>
  <div
    class="tox"
  >
    <div
      class="tox-expandable-box"
    >
      <div
        class="tox-expandable-box__content tox-expandable-box__content--overflowing"
        style="max-height: 80px;"
      >
        <div
          style="height: 200px;"
        >
          Hello world
        </div>
      </div>
      <button
        class="tox-expandable-box__toggle-button"
        type="button"
      >
        <span
          class="tox-icon"
        >
          Icon: chevron-down
        </span>
        Expand
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`browser.ExpandableBoxTest > Should be able to set maxHeight, collapseText and expandText > 1. Content fits within maxHeight 1`] = `
<DocumentFragment>
  <div
    class="tox"
  >
    <div
      class="tox-expandable-box"
    >
      <div
        class="tox-expandable-box__content tox-expandable-box__content--overflowing"
        style="max-height: 100px;"
      >
        <div
          style="height: 150px;"
        >
          Hello world
        </div>
      </div>
      <button
        class="tox-expandable-box__toggle-button"
        type="button"
      >
        <span
          class="tox-icon"
        >
          Icon: chevron-down
        </span>
        Show more
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`browser.ExpandableBoxTest > Should be able to set maxHeight, collapseText and expandText > 2. After show mode click 1`] = `
<DocumentFragment>
  <div
    class="tox"
  >
    <div
      class="tox-expandable-box"
    >
      <div
        class="tox-expandable-box__content tox-expandable-box__content--expanded"
        style="max-height: none;"
      >
        <div
          style="height: 150px;"
        >
          Hello world
        </div>
      </div>
      <button
        class="tox-expandable-box__toggle-button"
        type="button"
      >
        <span
          class="tox-icon"
        >
          Icon: chevron-up
        </span>
        Show less
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`browser.ExpandableBoxTest > Should not render expand if the content fits within maxHeight > Content fits within maxHeight 1`] = `
<DocumentFragment>
  <div
    class="tox"
  >
    <div
      class="tox-expandable-box"
    >
      <div
        class="tox-expandable-box__content"
        style="max-height: 80px;"
      >
        <div
          style="height: 50px;"
        >
          Hello world
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
